File: cpp/benchmarks/benchmark.cpp
/**
 * salman 40 - Enterprise C++ Benchmarks
 * Generated: 2025-08-28T10:28:26.240Z
 */

#include "averox_crypto.h"
#include <chrono>
#include <iostream>
#include <vector>
#include <thread>
#include <future>
#include <random>
#include <fstream>

class CppBenchmarkSuite {
private:
    struct BenchmarkResults {
        double encryptionOpsPerSecond;
        double decryptionOpsPerSecond;
        double throughputMBps;
        double averageLatencyMs;
        size_t memoryUsageKB;
    };
    
    std::map<std::string, BenchmarkResults> results_;

public:
    void runAllBenchmarks() {
        std::cout << "🚀 Starting C++ Enterprise Benchmarks..." << std::endl;
        
        benchmarkEncryptionPerformance();
        benchmarkLargeDataOperations();
        benchmarkConcurrentOperations();
        benchmarkMemoryUsage();
        benchmarkAlgorithmComparison();
        
        generateReport();
    }

private:
    void benchmarkEncryptionPerformance() {
        std::cout << "📊 Benchmarking encryption performance..." << std::endl;
        
        auto masterKey = generateRandomKey(32);
        AveroxCrypto::Config config;
        config.enableMetrics = true;
        
        AveroxCrypto::AveroxCrypto crypto(masterKey, config);
        
        const std::string testData(1024, 'x'); // 1KB test data
        const int iterations = 10000;
        
        // Benchmark encryption
        auto startTime = std::chrono::high_resolution_clock::now();
        
        for (int i = 0; i < iterations; ++i) {
            auto encrypted = crypto.encrypt(testData);
            auto decrypted = crypto.decrypt(encrypted);
        }
        
        auto endTime = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);
        
        BenchmarkResults result;
        result.encryptionOpsPerSecond = (iterations * 2.0) / (duration.count() / 1000000.0);
        result.throughputMBps = (testData.size() * iterations * 2.0) / (duration.count() / 1000000.0) / 1024 / 1024;
        result.averageLatencyMs = (duration.count() / 1000.0) / (iterations * 2);
        
        results_["encryption_performance"] = result;
    }
    
    void benchmarkLargeDataOperations() {
        std::cout << "📈 Benchmarking large data operations..." << std::endl;
        
        auto masterKey = generateRandomKey(32);
        AveroxCrypto::AveroxCrypto crypto(masterKey);
        
        std::vector<size_t> dataSizes = {1024*1024, 10*1024*1024, 100*1024*1024}; // 1MB, 10MB, 100MB
        
        for (size_t size : dataSizes) {
            std::string largeData(size, 'x');
            
            auto startTime = std::chrono::high_resolution_clock::now();
            auto encrypted = crypto.encrypt(largeData);
            auto encryptionTime = std::chrono::high_resolution_clock::now();
            auto decrypted = crypto.decrypt(encrypted);
            auto endTime = std::chrono::high_resolution_clock::now();
            
            auto encDuration = std::chrono::duration_cast<std::chrono::milliseconds>(encryptionTime - startTime);
            auto decDuration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - encryptionTime);
            auto totalDuration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
            
            BenchmarkResults result;
            result.throughputMBps = (size * 2.0) / (totalDuration.count() / 1000.0) / 1024 / 1024;
            
            std::string sizeLabel = std::to_string(size / 1024 / 1024) + "MB";
            results_["large_data_" + sizeLabel] = result;
        }
    }
    
    void benchmarkConcurrentOperations() {
        std::cout << "🔄 Benchmarking concurrent operations..." << std::endl;
        
        auto masterKey = generateRandomKey(32);
        const std::string testData = "concurrent test data";
        const int numThreads = std::thread::hardware_concurrency();
        const int operationsPerThread = 1000;
        
        auto startTime = std::chrono::high_resolution_clock::now();
        
        std::vector<std::future<bool>> futures;
        
        for (int i = 0; i < numThreads; ++i) {
            futures.push_back(std::async(std::launch::async, [&masterKey, &testData, operationsPerThread]() {
                AveroxCrypto::AveroxCrypto crypto(masterKey);
                
                for (int j = 0; j < operationsPerThread; ++j) {
                    auto encrypted = crypto.encrypt(testData);
                    auto decrypted = crypto.decrypt(encrypted);
                    if (decrypted != testData) return false;
                }
                return true;
            }));
        }
        
        int successCount = 0;
        for (auto& future : futures) {
            if (future.get()) successCount++;
        }
        
        auto endTime = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
        
        BenchmarkResults result;
        result.encryptionOpsPerSecond = (numThreads * operationsPerThread * 2.0) / (duration.count() / 1000.0);
        
        results_["concurrent_operations"] = result;
        
        std::cout << "Concurrent test: " << successCount << "/" << numThreads << " threads succeeded" << std::endl;
    }
    
    void benchmarkMemoryUsage() {
        std::cout << "💾 Benchmarking memory usage..." << std::endl;
        
        std::vector<std::unique_ptr<AveroxCrypto::AveroxCrypto>> instances;
        
        for (int i = 0; i < 1000; ++i) {
            auto masterKey = generateRandomKey(32);
            instances.push_back(std::make_unique<AveroxCrypto::AveroxCrypto>(masterKey));
        }
        
        // Simulate memory usage calculation
        BenchmarkResults result;
        result.memoryUsageKB = instances.size() * 2; // Approximate 2KB per instance
        
        results_["memory_usage"] = result;
    }
    
    void benchmarkAlgorithmComparison() {
        std::cout << "🔐 Comparing encryption algorithms..." << std::endl;
        
        auto masterKey = generateRandomKey(32);
        AveroxCrypto::AveroxCrypto crypto(masterKey);
        
        const std::string testData(10240, 'x'); // 10KB test data
        const int iterations = 1000;
        
        std::vector<std::string> algorithms = {"aes-256-gcm", "chacha20-poly1305"};
        
        for (const auto& algorithm : algorithms) {
            auto startTime = std::chrono::high_resolution_clock::now();
            
            for (int i = 0; i < iterations; ++i) {
                auto encrypted = crypto.encrypt(testData, "", algorithm);
                auto decrypted = crypto.decrypt(encrypted);
            }
            
            auto endTime = std::chrono::high_resolution_clock::now();
            auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);
            
            BenchmarkResults result;
            result.encryptionOpsPerSecond = (iterations * 2.0) / (duration.count() / 1000000.0);
            result.throughputMBps = (testData.size() * iterations * 2.0) / (duration.count() / 1000000.0) / 1024 / 1024;
            
            results_[algorithm + "_performance"] = result;
        }
    }
    
    void generateReport() {
        std::cout << "\n📋 C++ Enterprise Benchmark Report" << std::endl;
        std::cout << "===================================" << std::endl;
        
        std::ofstream reportFile("cpp-benchmark-results.json");
        reportFile << "{\n";
        
        bool first = true;
        for (const auto& [test, result] : results_) {
            if (!first) reportFile << ",\n";
            first = false;
            
            std::cout << test << ":" << std::endl;
            if (result.encryptionOpsPerSecond > 0) {
                std::cout << "  Operations/sec: " << static_cast<int>(result.encryptionOpsPerSecond) << std::endl;
            }
            if (result.throughputMBps > 0) {
                std::cout << "  Throughput MB/s: " << result.throughputMBps << std::endl;
            }
            if (result.memoryUsageKB > 0) {
                std::cout << "  Memory Usage KB: " << result.memoryUsageKB << std::endl;
            }
            
            reportFile << "  \"" << test << "\": {\n";
            reportFile << "    \"ops_per_second\": " << result.encryptionOpsPerSecond << ",\n";
            reportFile << "    \"throughput_mbps\": " << result.throughputMBps << ",\n";
            reportFile << "    \"memory_usage_kb\": " << result.memoryUsageKB << "\n";
            reportFile << "  }";
        }
        
        reportFile << "\n}";
        reportFile.close();
        
        std::cout << "\n💾 Results saved to cpp-benchmark-results.json" << std::endl;
    }
    
    std::vector<uint8_t> generateRandomKey(size_t size) {
        std::vector<uint8_t> key(size);
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, 255);
        
        for (size_t i = 0; i < size; ++i) {
            key[i] = static_cast<uint8_t>(dis(gen));
        }
        
        return key;
    }
};

int main() {
    CppBenchmarkSuite suite;
    suite.runAllBenchmarks();
    return 0;
}

File: python/benchmarks/benchmark.py
"""
salman 40 - Enterprise Python Performance Benchmarks
Generated: 2025-08-28T10:28:26.243Z
"""

import asyncio
import time
import statistics
import psutil
import concurrent.futures
from averox_crypto import AveroxCrypto, CryptoUtils
from async_crypto import AsyncAveroxCrypto, PerformanceMonitor

class PythonBenchmarkSuite:
    def __init__(self):
        self.results = {}
        self.monitor = PerformanceMonitor()
    
    async def run_all_benchmarks(self):
        """Run comprehensive benchmark suite"""
        print("🚀 Starting Python Enterprise Benchmarks...")
        
        await self.benchmark_sync_vs_async()
        await self.benchmark_encryption_algorithms()
        await self.benchmark_large_data()
        await self.benchmark_concurrent_operations()
        self.benchmark_memory_usage()
        
        self.generate_report()
    
    async def benchmark_sync_vs_async(self):
        """Compare sync vs async performance"""
        print("⚡ Benchmarking sync vs async...")
        
        master_key = CryptoUtils.generate_master_key()
        sync_crypto = AveroxCrypto(master_key)
        async_crypto = AsyncAveroxCrypto(master_key)
        
        test_data = "benchmark test data " * 100
        iterations = 1000
        
        # Sync benchmark
        sync_start = time.time()
        for _ in range(iterations):
            encrypted = sync_crypto.encrypt(test_data)
            sync_crypto.decrypt(encrypted)
        sync_time = time.time() - sync_start
        
        # Async benchmark
        async_start = time.time()
        tasks = []
        for _ in range(iterations):
            task = self._async_encrypt_decrypt(async_crypto, test_data)
            tasks.append(task)
        await asyncio.gather(*tasks)
        async_time = time.time() - async_start
        
        self.results['sync_vs_async'] = {
            'sync_ops_per_second': (iterations * 2 / sync_time),
            'async_ops_per_second': (iterations * 2 / async_time),
            'performance_gain': f"{((sync_time / async_time - 1) * 100):.1f}%"
        }
        
        sync_crypto.destroy()
        async_crypto.destroy()
    
    async def _async_encrypt_decrypt(self, crypto, data):
        """Helper for async encrypt/decrypt"""
        encrypted = await crypto.encrypt_async(data)
        return await crypto.decrypt_async(encrypted)
    
    async def benchmark_encryption_algorithms(self):
        """Benchmark different encryption algorithms"""
        print("🔐 Benchmarking encryption algorithms...")
        
        master_key = CryptoUtils.generate_master_key()
        crypto = AveroxCrypto(master_key)
        
        test_data = "x" * 10240  # 10KB
        iterations = 1000
        
        algorithms = ['aes-256-gcm', 'chacha20-poly1305']
        
        for algorithm in algorithms:
            start_time = time.time()
            for _ in range(iterations):
                encrypted = crypto.encrypt(test_data, algorithm=algorithm)
                crypto.decrypt(encrypted)
            elapsed = time.time() - start_time
            
            self.results[f'{algorithm}_performance'] = {
                'ops_per_second': (iterations * 2 / elapsed),
                'throughput_mbps': ((len(test_data) * iterations * 2) / elapsed / 1024 / 1024),
                'avg_latency_ms': (elapsed / iterations / 2 * 1000)
            }
        
        crypto.destroy()
    
    async def benchmark_large_data(self):
        """Benchmark large data encryption"""
        print("📊 Benchmarking large data operations...")
        
        master_key = CryptoUtils.generate_master_key()
        crypto = AsyncAveroxCrypto(master_key)
        
        sizes = [1024*1024, 10*1024*1024, 100*1024*1024]  # 1MB, 10MB, 100MB
        
        for size in sizes:
            data = "x" * size
            size_label = f"{size // 1024 // 1024}MB"
            
            start_time = time.time()
            encrypted = await crypto.encrypt_async(data)
            encryption_time = time.time() - start_time
            
            start_time = time.time()
            await crypto.decrypt_async(encrypted)
            decryption_time = time.time() - start_time
            
            self.results[f'large_data_{size_label}'] = {
                'encryption_time_s': encryption_time,
                'decryption_time_s': decryption_time,
                'total_time_s': encryption_time + decryption_time,
                'throughput_mbps': (size / (encryption_time + decryption_time) / 1024 / 1024)
            }
        
        crypto.destroy()
    
    async def benchmark_concurrent_operations(self):
        """Benchmark concurrent crypto operations"""
        print("🔄 Benchmarking concurrent operations...")
        
        master_key = CryptoUtils.generate_master_key()
        crypto = AsyncAveroxCrypto(master_key)
        
        concurrency_levels = [10, 50, 100, 200]
        test_data = "concurrent test data"
        
        for concurrency in concurrency_levels:
            start_time = time.time()
            
            tasks = []
            for _ in range(concurrency):
                task = self._async_encrypt_decrypt(crypto, test_data)
                tasks.append(task)
            
            results = await asyncio.gather(*tasks, return_exceptions=True)
            elapsed = time.time() - start_time
            
            successful = sum(1 for r in results if not isinstance(r, Exception))
            
            self.results[f'concurrent_{concurrency}'] = {
                'total_operations': concurrency,
                'successful': successful,
                'failed': concurrency - successful,
                'ops_per_second': (successful / elapsed),
                'success_rate': f"{(successful / concurrency * 100):.1f}%"
            }
        
        crypto.destroy()
    
    def benchmark_memory_usage(self):
        """Benchmark memory usage patterns"""
        print("💾 Benchmarking memory usage...")
        
        process = psutil.Process()
        initial_memory = process.memory_info().rss / 1024 / 1024  # MB
        
        instances = []
        for _ in range(1000):
            master_key = CryptoUtils.generate_master_key()
            instances.append(AveroxCrypto(master_key))
        
        peak_memory = process.memory_info().rss / 1024 / 1024  # MB
        
        # Cleanup
        for instance in instances:
            instance.destroy()
        
        final_memory = process.memory_info().rss / 1024 / 1024  # MB
        
        self.results['memory_usage'] = {
            'initial_memory_mb': initial_memory,
            'peak_memory_mb': peak_memory,
            'final_memory_mb': final_memory,
            'memory_per_instance_kb': ((peak_memory - initial_memory) * 1024 / 1000)
        }
    
    def generate_report(self):
        """Generate comprehensive benchmark report"""
        print("\n📋 Python Enterprise Benchmark Report")
        print("=" * 40)
        
        import json
        print(json.dumps(self.results, indent=2))
        
        with open('python-benchmark-results.json', 'w') as f:
            json.dump(self.results, f, indent=2)
        
        print("\n💾 Results saved to python-benchmark-results.json")

# CLI runner
async def main():
    suite = PythonBenchmarkSuite()
    await suite.run_all_benchmarks()

if __name__ == "__main__":
    asyncio.run(main())